general/echo/kmdf/exe/src/main.rs

```diff
@@ -0,0 +1,555 @@
+#![deny(clippy::unnecessary_safety_doc)]
+#![deny(rustdoc::broken_intra_doc_links)]
+#![deny(rustdoc::private_intra_doc_links)]
+#![deny(rustdoc::missing_crate_level_docs)]
+#![deny(rustdoc::invalid_codeblock_attributes)]
+#![deny(rustdoc::invalid_html_tags)]
+#![deny(rustdoc::invalid_rust_codeblocks)]
+#![deny(rustdoc::bare_urls)]
+#![deny(rustdoc::unescaped_backticks)]
+#![deny(rustdoc::redundant_explicit_links)]
+
+use std::{env, error::Error, ffi::OsString, os::windows::prelude::*, sync::RwLock, thread};
+
+use once_cell::sync::Lazy;
+use uuid::{uuid, Uuid};
+use windows_sys::Win32::{
+    Devices::DeviceAndDriverInstallation,
+    Foundation::{CloseHandle, GetLastError, BOOL, ERROR_IO_PENDING, HANDLE, INVALID_HANDLE_VALUE},
+    Storage::FileSystem::{
+        CreateFileW,
+        ReadFile,
+        WriteFile,
+        FILE_FLAG_OVERLAPPED,
+        FILE_GENERIC_READ,
+        FILE_GENERIC_WRITE,
+        FILE_SHARE_READ,
+        FILE_SHARE_WRITE,
+        OPEN_EXISTING,
+    },
+    System::{
+        Threading::INFINITE,
+        IO::{CreateIoCompletionPort, GetQueuedCompletionStatus, OVERLAPPED, OVERLAPPED_0},
+    },
+};
+
+#[derive(Default, Debug)]
+struct Globals {
+    perform_async_io: bool,
+    limited_loops: bool,
+    async_io_loops_num: usize,
+    device_path: String,
+}
+
+static GLOBAL_DATA: Lazy<RwLock<Globals>> = Lazy::new(|| RwLock::new(Globals::default()));
+static GUID_DEVINTERFACE_ECHO: Uuid = uuid!("CDC35B6E-0BE4-4936-BF5F-5537380A7C1A");
+static READER_TYPE: u32 = 1;
+static WRITER_TYPE: u32 = 2;
+static NUM_ASYNCH_IO: usize = 100;
+static BUFFER_SIZE: usize = 40 * 1024;
+
+fn main() -> Result<(), Box<dyn Error>> {
+    let argument_vector: Vec<String> = env::args().collect();
+    let argument_count = argument_vector.len();
+
+    if argument_count > 1 {
+        if argument_vector[1] == "-Async" {
+            let mut globals = GLOBAL_DATA.write()?;
+            globals.perform_async_io = true;
+            if argument_count > 2 {
+                globals.async_io_loops_num = argument_vector[2].parse::<usize>()?;
+                globals.limited_loops = true;
+            } else {
+                globals.limited_loops = false;
+            }
+        } else {
+            eprintln!(
+                r##"
+Usage:
+    Echoapp.exe         --- Send single write and read request synchronously
+    Echoapp.exe -Async  --- Send reads and writes asynchronously without terminating
+    Echoapp.exe -Async <number> --- Send <number> reads and writes asynchronously
+Exit the app anytime by pressing Ctrl-C
+"##
+            );
+            return Err("Invalid Args".into());
+        }
+    }
+
+    get_device_path(&GUID_DEVINTERFACE_ECHO)?;
+
+    let globals = GLOBAL_DATA.read()?;
+    println!("DevicePath: {}", globals.device_path);
+    let mut path_vec = globals.device_path.encode_utf16().collect::<Vec<_>>();
+    let perform_async_io = globals.perform_async_io;
+    drop(globals);
```

Comment:
although the lint does suggest to use drop, its more idiomatic to have it a drop by putting the block in braces